checkbox-ng/plainbox/impl/session/assistant.py

```diff
@@ -1704,20 +1704,31 @@ def finalize_session(self) -> None:
         them.
         """
         UsageExpectation.of(self).enforce()
-        if SessionMetaData.FLAG_INCOMPLETE not in self._metadata.flags:
+        finalizable_flags = [
+            SessionMetaData.FLAG_INCOMPLETE,
+            SessionMetaData.FLAG_BOOTSTRAPPING,
+        ]
+
+        if all(
+            [flag not in self._metadata.flags for flag in finalizable_flags]
+        ):
             _logger.info(
                 "finalize_session called for already finalized" " session: %s",
                 self._manager.storage.id,
             )
             # leave the same usage expectations
             return
-        if SessionMetaData.FLAG_SUBMITTED not in self._metadata.flags:
+        submitted = SessionMetaData.FLAG_SUBMITTED in self._metadata.flags
+        fresh = SessionMetaData.FLAG_BOOTSTRAPPING in self._metadata.flags
+        if not fresh and not submitted:
```

Comment:
Minor: it is slightly clearer to me `if incomplete and not submitted:`, here the logic is reversed

Reply:
Incomplete and bootstrapping flags are not the same. This is why I introduced the concept of "fresh".
For instance the session that got started but did not go through bootstrapping yet wouldn't have the bootstrapping flag (incomplete neither), and would be totally disposable, so we don't have to warn about that.

Reply:
what about `if completed_bootstrapping and not submitted`? my comment was more on the fact that I find hard to reason on what `not condition and not other_condition` mean, I find it simpler to grasp when at least one is positive

Reply:
Interesting. So my thinking behind this is:

establish facts
write condition based on that fact

And I find it better for facts to not be written in the inverse logic especially when they describe some existing state.


In other words:
```
good_weather = sunny and temp > 22.0
if not good_weather:
    stay_home()
```

vs

```
bad_weather = not sunny or temp <= 22.0
if bad_weather:
    stay_home()
```


The other thing is that the "fresh" is a moving target, for now it means started but not bootstrapped. But this may change in the future requiring more than one change. Of course now it's so local that this doesn't have big value.

Also the session may be interrupted _while_ bootstrapping, making the "completed_bootstrapping" a bit of a lie.

Reply:
Ok then, this is clearly personal preference. I would at least try a different name, fresh is a bit... funny, I would not know what it is by just reading that line. Something like `relevant`, `started_bootstrapping`, `non_empty`. Also, maybe a comment to explain what it is above it, I did not get it at first, I think I do now, something like `ignore session that did not even reach bootsrapping`

Reply:
Fixed in a411e4e07